Excited groups no longer breakdown during certain gas reactions.

Pickle-Coding opened the pull request:
## About The Pull Request
Excited groups no longer breakdown during certain gas reactions that are considered too volatile without necessarily being a fire.
## Why It's Good For The Game
This stops shit like antinoblium from breaking down and equalising with a large chunk of the station instantly. Should make that feel more like goop spreading.
## Changelog
:cl:
balance: Some gas reactions will prevent excited groups from breaking down.
/:cl:

Pickle-Coding committed:
Add VOLATILE_REACTION flag.

Adds the VOLATILE_REACTION bitflag define for reaction bitflags. Does nothing for now. Gas reactions that are considered volatile will have this bitflag. Changes some of the gas reactions method of returning to add VOLATILE_REACTION and REACTING as bitflags.

VOLATILE_REACTION is intended for preventing excited groups from breaking down in a future commit.

Pickle-Coding committed:
Merge branch 'master' of https://github.com/Pickle-Coding/tgstation into volatile-reactions

Pickle-Coding committed:
Excited groups no longer breakdown during a volatile reaction.

Also apply VOLATILE_REACTION shit to antinoblium_replication.

I don't know what happened but maybe I commited while it was compiling and it fucked shit up with tracking files that shouldn't be tracked. It was extremely fucked up.

Pickle-Coding committed:
Fixes erroneoijoeus space.

???

Xander3359 commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -259,13 +261,14 @@
 	)
 
 /datum/gas_reaction/h2fire/react(datum/gas_mixture/air, datum/holder)
+	. = VOLATILE_REACTION
```
Rather than copypaste this on every gas.
It's probably better to have a `var/gas_flags` on `/datum/gas_reaction`
And then assign volatility on parent call

Pickle-Coding replied:
Gas reactions don't call parent and some of these volatile flags are conditional. Not sure it would be worth it.

Xander3359 replied:
ok but Im saying you could simply add `. = ..()` and then in the parent call just change the behaviour based on a flag you add to each subtype.

LemonInTheDark commented on code/controllers/subsystem/air.dm:
```diff
@@ -391,7 +391,7 @@ SUBSYSTEM_DEF(air)
 		EG.breakdown_cooldown++
 		EG.dismantle_cooldown++
-		if(EG.breakdown_cooldown >= EXCITED_GROUP_BREAKDOWN_CYCLES)
+		if(EG.breakdown_cooldown >= EXCITED_GROUP_BREAKDOWN_CYCLES && !(EG.turf_reactions & (REACTING | VOLATILE_REACTION)))
```
what do you intend this check to do? also I am not sure if this would lead to breakdown -> dismantle in one tick

Pickle-Coding replied:
Fail if there is a reaction that is REACTING and is a VOLATILE_REACTION.

LemonInTheDark commented on code/__DEFINES/atmospherics/atmos_core.dm:
```diff
@@ -68,14 +68,16 @@
+// Return values for reactions (bitflags).
+/// The gas mixture is not reacting.
 #define NO_REACTION 0
-///The gas mixture is reacting
-#define REACTING 1
-///The gas mixture is able to stop all reactions
-#define STOP_REACTIONS 2
+/// The gas mixture is reacting.
+#define REACTING (1 << 0)
+/// The gas mixture is able to stop all reactions.
+#define STOP_REACTIONS (1 << 1)
+/// The gas mixture has the requirements to start a volatile reaction.
+#define VOLATILE_REACTION (1 << 2)
```
formatting should be locally consistent, which it already isn't but this is just a third type

LemonInTheDark commented on code/__DEFINES/atmospherics/atmos_core.dm:
```diff
@@ -68,14 +68,16 @@
+// REACTIONS
+// Return values for reactions (bitflags).
+/// The gas mixture is not reacting.
 #define NO_REACTION 0
```
NONE?

Pickle-Coding replied:
I forgot about this. I made this PR months ago but only decided to finish the last 1% today. I don't fully remember the justification. Don't remember if anything expects NO_REACTION to be falsy. I also assume 0 is just 00000000000000000000000 for the bitflags. Maybe it would make more sense to remove NO_REACTION.

LemonInTheDark reviewed (changes requested):
#### Ones that don't change much behavior wise
```
Plasmafire
trit fire
freon fire
```
#### Ones I otherwise agree with
```
proto nitrate balls/rads
h2 fire
```
 #### Ones I'm confused about
```
nitrous decomp
nob formation
halon o2 purging
```
I don't really get why we're doing this, these seem... harmless to me, or at least ok to spread out instantly, it doesn't seem like it would feel weird like fire does
#### Misc
```
Water
```
I get where you're coming from with this. It looks weird for it to suddenly jump out. I'm kind of concerned cause the reaction is built to run forever until all the water's gone, and this would lock up breakdown for a while. I think this is sort of just a problem with all visible gas, and less with water? idk.
``` 
antinob replication
```
Because this basically runs infinitely I think this would end up locking itself up badly, anywhere it is there antinob the reaction's gonna be running it'll never settle, and it won't be able to effectively "move". It's possible the reaction would overcome this, but I'm pretty sure it would basically never be able to settle out and clean itself up, ever.

Pickle-Coding committed:
Nitrous oxide decomposition no longer has VOLATILE_REACTION.

Pickle-Coding committed:
VOLATILE_REACTION only if the reaction reacts successfully.

The VOLATILE_REACTION flag only gets applied when the gas reactions are successful. This fixes some edgecases of preventing breakdown during gas reactions that can be volatile but aren't reacting properly while another boring gas reaction reacts.

Pickle-Coding committed:
Better defines formatting.

Spaces the comments to be locally consistent. Uses NONE instead of 0 for NO_REACTION.

LemonInTheDark commented on code/controllers/subsystem/air.dm:
```diff
@@ -391,7 +391,7 @@ SUBSYSTEM_DEF(air)
 		EG.dismantle_cooldown++
-		if(EG.breakdown_cooldown >= EXCITED_GROUP_BREAKDOWN_CYCLES)
+		if(EG.breakdown_cooldown >= EXCITED_GROUP_BREAKDOWN_CYCLES && !(EG.turf_reactions & VOLATILE_REACTION))
 			EG.self_breakdown(poke_turfs = TRUE)
```
```suggestion
			EG.self_breakdown(poke_turfs = TRUE)
			EG.dismantle_cooldown = EXCITED_GROUP_BREAKDOWN_CYCLES
```
Otherwise, you risk instantly dismantling in a non stable state.

LemonInTheDark commented:
> Nitrous decomposition is fairly exothermic, though since it already has a fairly high temperature requirement it actually doesn't matter so I agree it shouldn't be volatile. Noblium formation releases an absurd amount of energy so it's reasonable to assume it's volatile, though the reaction will be quick either ways so it might not change much. Halon combustion releases resin so it would be silly if that equalised rapidly.

I'm fine with this

> With water it only counts as volatile if it successfuly creates the ice, so that will only occur while it's actively spreading to new tiles. Antinoblium replication only returns with REACTING if it is actually replicating. Its speed is based purely on its own moles so it's not a slow reaction to clear up the remaining gases.

Similar RE: water. Antinob concerns me because it's basically always reacting, so in a spreading situation I think it could end up stalling out really really easily, which I don't like.

Pickle-Coding committed:
Set dismantle_cooldown to EXCITED_GROUP_BREAKDOWN_CYCLES after a self_breakdown.

Don't want to instantly dismantle in an unstable state.

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

LemonInTheDark commented on code/controllers/subsystem/air.dm:
```diff
@@ -391,8 +391,9 @@ SUBSYSTEM_DEF(air)
-		if(EG.breakdown_cooldown >= EXCITED_GROUP_BREAKDOWN_CYCLES)
+		if(EG.breakdown_cooldown >= EXCITED_GROUP_BREAKDOWN_CYCLES && !(EG.turf_reactions & VOLATILE_REACTION))
 			EG.self_breakdown(poke_turfs = TRUE)
+			EG.dismantle_cooldown = EXCITED_GROUP_BREAKDOWN_CYCLES
```
```suggestion
			EG.dismantle_cooldown = min(EG.dismantle_cooldown, EXCITED_GROUP_BREAKDOWN_CYCLES) // Can be delayed
```
sorry I'm stupid

Pickle-Coding replied:
I think this might be causing problems with antinoblium replication. I'll investigate it further.

Pickle-Coding replied:
I think this stops excited groups from being able to dismantle.

Pickle-Coding committed:
Lemon was stupid.

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

Pickle-Coding commented:
I think for water I could possibly let it breakdown if it didn't spread to any new turfs. I suppose I could see how long antinoblium would practically spread.

Pickle-Coding committed:
Only volatile if it is applying to unaffected turfs.

Water vapour deposition is only volatile if it is freezing turfs that are not already frozen. Halon is only volatile if it is actually creating resin.

Pickle-Coding committed:
Fix logic for delaying dismantle_cooldown.

Don't increase dismantle cooldown when breakdown gets delayed by a volatile reaction. This prevents dismantling in an unstable state.
